### ISSUE TITLE
1920x1080 project on a 4k Screen lag

### ISSUE THREAD

**Issue by colinharvie** [renpy/renpy#5344]
So I've got a game I've made, and it runs butter smooth on HD. 
4K is another matter. It starts chugging real hard.

Is there anyway to just.... I guess render it as if it was 1920x1080, then just scale up the final image? I don't understand everything behind the scenes with rendering, but it does seem like it's trying to render everything at 4k. 

I understand that it won't be sharp, but it still would look acceptable, especially since my assets are only made for HD screens anyway.

(Part of hte problem is that I'm also using Live2D models, which seem to multiply the problem and lag like heck on a 4k screen.) 

I can use config.adjust_vew_size to limit the size, which solves the lagging issue, but then everything is surrounded by black.

Is there some way currently in Ren'py to deal with this, or could it be implemented? Something where it just scales up the 1920x1080 render after the fact?

Here's the config.adjust_view_size function I use, if needed:
```
init python:
    def adjust_view_size(width,height):
        screen_width = config.screen_width
        screen_height = config.screen_height
        w = width if width < screen_width else screen_width
        h = height if height < screen_height else screen_height
        return (w,h)

    config.adjust_view_size = adjust_view_size 
```

**Comment by renpytom** [comment]
You can set the games size to (1920, 1080), then use config.physical_size to have it default to something bigger, if possible. But if your Live2D uses huge textures, those will still be loaded and used.

**Comment by colinharvie** [comment]
I'm not sure if I didn't explain well, or if I'm misunderstanding, but this doesn't seem to make a difference.

I had this previously:
```
define config.physical_height = 1080
define config.physical_width = 1920
define config.screen_height = 1080
define config.screen_width = 1920
```
I changed it to
```
define config.physical_height = 2160
define config.physical_width = 3840
define config.screen_height = 1080
define config.screen_width = 1920
```

I don't know if I didn't explain correctly,  but my issue is not that I want the physical window size to be 4k. I just want the game to be able to run if the user chooses to play it on a 4k monitor, in full screen mode.

I understand that the textures (which are large), will be loaded either way, HOWEVER I have used incredibly scaled down textures (under 200px) and that did not make a difference. When the game runs in full screen mode on a 4k screen, even if the virtual screen size is set to 1920x1080, the game will still render at 4k. I can visibly see the sharper texture of text or images that were scaled down with a Transform, and are now next to HD assets, on a 4k screen. 

I'm wondering if it's possible to have the game render at 1920x1080, as if that were the physical screen size, and only then scale it up to fit on a 4k screen, instead of rendering at the physical screen size. 

The Live2D runs perfectly smooth at the 1920x1080, regardless of my textures.

**Comment by colinharvie** [comment]
Some further information regarding the Live2d, in case it helps.
The lag seems to be caused by the rendering of the alpha/clipping masks. 
My work around for this (at 1920x1080) was to set the Zoom on the Live2D object to 0.125, and then wrapping the whole thing in a Transform(zoom=8). This has the effect of rendering the clipping masks at 1/8th the resolution, but the rest of the model still rendered at the computed 1.0 scale in the end. This can cause some fuzzing at the edges of the clipping masks, unfortunately.

The reason why I say it seems like everything is still being rendered larger "behind-the-scenes" when it goes to 4k, is that that I can get it to run smoothly at 4k IF I multiply the Zoom on the Live2d object and on it's wrapping transform to account for the screen change (0.5 and 2 respectively)

It runs smooth again, but the fuzzing on the model is more pronounced.

Typing all this out, I've come up with a work around, which would be to dynamically change that scaling factor on wrapper and on the Live2D class. I will attempt that, but it feels like this is an overcomplicated (and probably not great) solution.

**Comment by renpytom** [comment]
Ren'Py does require your GPU be able to supply your screen with pixels. It will be slow if your hardware can't keep up with the fillrate your game requires, but the solution to that is to get hardware that supports your screen, rather than playing with scaling.

**Comment by colinharvie** [comment]
I was able to get my janky work around working. I think this is actually some sort of Live2D alpha mask issue in the end.

**Comment by renpytom** [comment]
I'm going to close this - there are some issues with Live2D performance that will be fixed in 8.2.1., and you seem to have done something to solve this.

**Comment by colinharvie** [comment]
Throwing this info here in case it's helpful:

It was definitely an issue with Live2D that I was having, but I think it's separate from whatever optimizations were done for 8.2.1 (I've tried the nightly, it did not seem to make a difference).

The way I sovled my problem was three-fold.
1. I super-optimized my Live2D Rigs, using only 4 or less clipping masks. It rather limits some of what I can do, but it was the only way I could get things to run smoothly.

I know from reading the Live2D SDK notes (though I'm far from savvy enough to understand the SDK), that the way Live2D handles masks can cause issues with more than 4 masks. That said, the performance in Ren'Py seems WAY worse (unusably so, and I have a fairly decent rig) than in other programs using Live2D models. Models that run smoothly both in the Live2D editor and other third party Live2D programs run perfectly smooth, but can be basically unusable in Ren'py. I don't know if this is just to do with Python or something else, but it's definitely related to the alpha masks / clipping.

2. Even with the optimized masks, the masks are still hard on the rig. One way I "cheated" my way through this was by setting the Zoom of the Live2D object to 0.25, and then wrapping it in a Transform(zoom=4). This had the effect of rendering the alpha mask at 1/4th the resolution, which was still good enough visually for me, but the rest of the model still rendered at 1x. (This is more obvious when really exagerating the scale difference, ie, 0.05x and 20x, becasue you can see the feathering caused by the lower resolution clipping masks) 

3. And the final thing I did, which specifically fixed my issue on 4k screens, was to have an update function that changed the Zoom level of the Live2D object AND the wrapping transform, based on the physical screen size. The root of my lag problem was caused because, upon scaling up to 4k, the Live2D masks (which are the source of the lag), would also scale up. To prevent this from happening, I compared the screen_size to the physical_size, if physical_size was larger, get the ratio, then adjust the Zoom of hte live2D object and the wrapping transform accordingly, so it was locked to a maximum.

Now things run butter smooth, but if that live2d alpha mask issue is something solvable... that'd be really cool, as being limited to 4 masks (that I keep as physically small as possible), really puts a damper on how advanced the rigs can be. I can open up another topic for that, though I was trying to wrap my head around the SDK (a very, VERY slow going process....) before I wrote anything and had a better idea of the problem.
